runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -8077,75 +8106,109 @@ static TR::Register *genCAS(TR::Node *node, TR::CodeGenerator *cg, TR::Register
    if (!casWithoutSync)
       generateInstruction(cg, TR::InstOpCode::sync, node);
-   generateTrg1ImmInstruction(cg, TR::InstOpCode::li, node, resultReg, 1);
+
+   if (!isExchange)
+      {
+      generateTrg1ImmInstruction(cg, TR::InstOpCode::li, node, resultReg, 1);
+      }
 
    node->setRegister(resultReg);
    return resultReg;
    }
 
-static TR::Register *VMinlineCompareAndSwap(TR::Node *node, TR::CodeGenerator *cg, bool isLong)
+static TR::Register *VMinlineCompareAndSetOrExchange(TR::Node *node, TR::CodeGenerator *cg, int32_t dataSize, bool isExchange)
    {
    TR::Compilation * comp = cg->comp();
    TR::Register *objReg, *offsetReg, *oldVReg, *newVReg, *resultReg, *cndReg;
-   TR::Node *firstChild, *secondChild, *thirdChild, *fourthChild, *fifthChild;
+   TR::Node *firstChild, *objNode, *offsetNode, *oldVNode, *newVNode;
    TR::RegisterDependencyConditions *conditions;
-   TR::LabelSymbol *doneLabel;
-   intptr_t offsetValue, oldValue;
+   TR::LabelSymbol *startLabel, *doneLabel;
+   int64_t offsetValue, oldValue;
    bool oldValueInReg = true, freeOffsetReg = false;
    TR_J9VMBase *fej9 = (TR_J9VMBase *) (cg->fe());
 
    firstChild = node->getFirstChild();
-   secondChild = node->getSecondChild();
-   thirdChild = node->getChild(2);
-   fourthChild = node->getChild(3);
-   fifthChild = node->getChild(4);
-   objReg = cg->evaluate(secondChild);
+   objNode = node->getSecondChild();
+   offsetNode = node->getChild(2);
+   oldVNode = node->getChild(3);
+   newVNode = node->getChild(4);
+
+   objReg = cg->evaluate(objNode);
 
    // VM helper chops off the value in 32bit, and we don't want the whole long value either
-   if (thirdChild->getOpCode().isLoadConst() && thirdChild->getRegister() == NULL && comp->target().is32Bit())
+   if (offsetNode->getOpCode().isLoadConst() && offsetNode->getRegister() == NULL && comp->target().is32Bit())
       {
-      offsetValue = thirdChild->getLongInt();
+      offsetValue = offsetNode->getLongInt();
       offsetReg = cg->allocateRegister();
       loadConstant(cg, node, (int32_t) offsetValue, offsetReg);
       freeOffsetReg = true;
       }
    else
       {
-      offsetReg = cg->evaluate(thirdChild);
+      offsetReg = cg->evaluate(offsetNode);
+
+      /* Assume that the offset is positive and not pathologically large (i.e., > 2^31). */
       if (comp->target().is32Bit())
          offsetReg = offsetReg->getLowOrder();
       }
 
-   if (fourthChild->getOpCode().isLoadConst() && fourthChild->getRegister() == NULL)
+   if (oldVNode->getOpCode().isLoadConst() && oldVNode->getRegister() == NULL)
       {
-      if (isLong)
-         oldValue = fourthChild->getLongInt();
-      else
-         oldValue = fourthChild->getInt();
-      if (oldValue >= LOWER_IMMED && oldValue <= UPPER_IMMED)
-         oldValueInReg = false;
+      switch (dataSize)
+         {
+         case 4:
+            oldValue = oldVNode->getInt();
+            if (oldValue >= LOWER_IMMED && oldValue <= UPPER_IMMED)
```

Comment:
this "if" block is expected to be common to all cases (even for future byte/short CAS support). why not move/common it to after-if?

Reply:
I can move it outside the switch block.